### ISSUE TITLE
Double processing

### ISSUE THREAD

**Issue by brianzhouzc** [BCCF-UBCO-AD/Orthanc-TMI#92]
I first noticed this bug when I'm working on the [feat-cstore](https://github.com/BCCF-UBCO-AD/Orthanc-TMI/tree/feat-cstore) branch. DICOM files are being processed twice when Orthanc receives them (currently only tested with WebUI).

DICOM files are being treated as two separate instances and are given two different UUIDs by Orthanc. However, only one instance of the file exists in the database after it's been processed, which is extremely strange.

This behavior still exists in the [develop](https://github.com/BCCF-UBCO-AD/Orthanc-TMI/tree/develop) branch. The file is being filtered twice:
```bash
orthanc-server  | W0121 09:09:04.139114 PluginsManager.cpp:168] Filter: compile new dicom buffer
...
orthanc-server  | W0121 09:09:04.140074 PluginsManager.cpp:168] Filter: compile new dicom buffer
...
orthanc-server  | writing file: /var/lib/orthanc/dicom/by-uuid/7a1/16/7a116e4e-df01-4ea0-bb37-2b8ee580983a.DCM
orthanc-server  | writing file: /var/lib/orthanc/dicom/by-uuid/88d/98/88d98971-a8d9-458a-8825-77695b97f734.DCM
orthanc-server  | WriteDicomFile: permissions set
orthanc-server  | WriteDicomFile: success
orthanc-server  | WriteDicomFile: permissions set
orthanc-server  | WriteDicomFile: success
```
But only one instance is written to the database and disk (88d98971-a8d9-458a-8825-77695b97f734). The other instance (7a116e4e-df01-4ea0-bb37-2b8ee580983a) are nowhere to be found in both database and on disk but the directory (/var/lib/orthanc/dicom/by-uuid/7a1/16/) is being generated. This implies the instance has been filtered as well, which is no good.

I suspect this has to do with files being uploaded via WebUI instead of being pushed to Orthanc via DICOM protocols. #53 will be a good starting point to investigate this problem. Maybe this is just a bug related to WebUI which we don't need to worry about

**Comment by brianzhouzc** [comment]
This issue only happens when duplicate files are uploaded. This happens with uploading directly with DICOM protocol as well.

**Comment by cppcooper** [comment]
I wonder if we could use the filter callback to detect full on duplicates and reject it there and then..

**Comment by cppcooper** [comment]
I still think the above is a good idea. But I was wondering about the Postgres side which brought me to:
https://www.postgresql.org/files/developer/concurrency.pdf
On page 31 it says:
> Actually, it’s not possible for two transactions to try to insert “the same row”
– any inserted row is a unique object by definition. So multiple insertions
can happen in parallel as far as PostgreSQL is concerned.
>
> The only way to get a conflict during insert is if you have a uniqueness
constraint (a unique index) on a table. If two concurrent transactions try to
insert rows having the same key value, then the second one will block until
the first one finishes. If the first transaction commits, the second one must
abort because of the uniqueness constraint; but if the first one aborts the
second one can proceed.

So I think this actually gives more credence to trying to detect file duplicates as early as possible.. Or prevent their root cause. We don't know what the root cause is yet, but presuming we can't figure it out.

What I'm thinking is:

1. Implement the `OrthancPluginIncomingDicomInstanceFilter`
    1. Anonymize file (maybe optional here?)
    2. Calculate md5
    3. Check if duplicates exist (either in memory or DB) and these need to be thread safe
    4. Save non-duplicates for storage, also needs to be thread safe since iii is probably coupled with this
    5. Reject duplicates (first come first kept, all else reject)
2. Take whatever gets through and save it to disk
    1. Retrieve the filtered file from memory
    2. Save it to disk
    3. Calculate filtered md5 (only if not already done)
    4. Save the uuid
3. After storing is complete update the checksum

We won't have the uuid during step 1. So we'll need to update the saved data once we have it.. or transfer that data and the md5 into a new container specifically for the checksum update to use.

All of our containers should be thread safe since this issue shows us orthanc runs itself in parallel when necessary.

**Comment by cppcooper** [comment]
I suppose we can close this with #113

**Comment by cppcooper** [comment]
Still happens, but isn't a problem so I am closing now.
